### PR TITLE
MRG: Unify covariance code in mne/decoding and mne/cov.py

### PR DESCRIPTION
**Pull request by larsoner**
I've been trying to use some decoding functions (`XDawn`, `CSP`) on rank-reduced SSS'ed MEG data. I'd like to be able to use the simple / naive `mne.cov.regularize` function for this, because it seems like I don't get consistent results from the automated regularizers (though maybe I'm just being dumb). Adding a `reg=callable` possibility seems like the easiest way, then I can do e.g.:
```
def reg(data):
    raw = mne.io.RawArray(data, epochs.info)
    cov = mne.compute_raw_covariance(raw, step=None)
    return mne.cov.regularize(cov, raw.info)['data']
```
So this fits my use case pretty easily, and seems like a pretty simple extension.

Todo:

- [x] add tests
- [x] fix docstring of CSP refs

@kingjr @alexandrebarachant @agramfort does this seem like a reasonable addition to you? If so, I'll update all other docstrings and add some tests.

### CODE REVIEW

**Comment by kingjr** [comment]
I think there is a couple options in pyriemann which may be a good API:
- `Xdawn(baseline_cov=your_manual_cov)` which allows you to build your
baseline covariance manually
- explicit shrinkage:

```python
pyriemann.estimation import Covariances, Shrinkage
make_pipeline(Covariances(), Shrinkage())
```

Note that the former option may lead to to non fully independence in case
of CV


Note that this approach can be
Don't we have the possibility to pass the covariance we want in.

**Comment by larsoner** [comment]
> - `Xdawn(baseline_cov=your_manual_cov)` 

Something like this already exists, at least in `XDawn`. The problem is that you can't do this for every cov that needs to be used by `XDawn` (or `CSP`)  -- some are calculated internally. So I don't think this option will work.

> - explicit shrinkage:

Wouldn't we need some sort of "reg as callable" to allow this anyway?

**Comment by larsoner** [comment]
Okay I've changed this PR to unify how covariances are handled in `mne.compute_covariance` and our decoding functions. I think this is cleaner than allowing a callable. (If we want to add that later, we can, and now if we add it to `mne/cov.py` we will have added it in `mne/dedoding`, too.)

Ready for review/merge from my end. @jona-sassenhagen you most recently hit a problem with this in `mne.decoding` so you might be interested in trying it. @kingjr you had some API thoughts, see if this makes sense if you have time.

@agramfort @dengemann this adds a couple of options to `mne.compute_covariance` to bring it to parity with those available in `mne.decoding`. Let me know what you think.

**Comment by dengemann** [comment]
@larsoner Yes I like this. Do we also allow for method params as we do in the old estimators? I did not see the documentation growing for the new estimators.

**Comment by larsoner** [comment]
> Do we also allow for method params as we do in the old estimators? I did not see the documentation growing for the new estimators.

Yep. It expands less than you might expect because I deduplicated the documentation. Read this more closely and it will probably make sense:

https://github.com/mne-tools/mne-python/pull/4808/files#diff-3e1bb52c1df3d57e5a84d9720d576712R614

File: mne/decoding/csp.py [hunk 1]
```diff
@@ -31,12 +31,13 @@ class CSP(TransformerMixin, BaseEstimator):
...
-        if float, shrinkage covariance is used (0 <= shrinkage <= 1).
-        if str, optimal shrinkage using Ledoit-Wolf Shrinkage ('ledoit_wolf')
-        or Oracle Approximating Shrinkage ('oas').
-    log : None | bool, defaults to None
+    reg : float | str | None (default None)
+        If not None (same as ``'empirical'``, default), allow
+        regularization for covariance estimation.
+        If float, shrinkage is used (0 <= shrinkage <= 1).
+        For str options, ``reg`` will be passed to ``method`` to
+        :func:`mne.compute_covariance`.
```
Comment (agramfort) [inline mne/decoding/csp.py]: there is no mention anymore to 'ledoit_wolf' or 'oas'

File: mne/decoding/csp.py [hunk 2]
```diff
@@ -680,19 +680,24 @@ class SPoC(CSP):
...
```
Comment (agramfort) [inline mne/decoding/csp.py]: same here

File: mne/decoding/csp.py (hunk 1 above)
Comment (larsoner) [reply mne/decoding/csp.py]: Backward compatibility. The former is in `compute_covariance` and the latter in `XDawn` / `CSP` in 0.15

Comment (larsoner) [reply mne/decoding/csp.py]: Oh I misread that. The idea is that the user should look at the doc of `compute_covariance`, I'll make it clearer

**Comment by larsoner** [comment]
Doc clarified

File: mne/decoding/csp.py [hunk 3]
```diff
@@ -86,22 +91,14 @@ class CSP(TransformerMixin, BaseEstimator):
     """
 
     def __init__(self, n_components=4, reg=None, log=None, cov_est="concat",
-                 transform_into='average_power', norm_trace=False):
+                 transform_into='average_power', norm_trace=False,
+                 method_params=None):
```
Comment (agramfort) [inline mne/decoding/csp.py]: method_params in a CSP class does not suggest it's related to cov estimation.

I would stick to the reg callable and expert users can do a functools.partial to set their custom params.

File: mne/decoding/csp.py [hunk 4]
```diff
@@ -232,7 +232,7 @@ def fit(self, X, y):
         eigen_vectors = eigen_vectors[:, ix]
 
         self.filters_ = eigen_vectors.T
-        self.patterns_ = linalg.pinv(eigen_vectors)
+        self.patterns_ = linalg.pinv2(eigen_vectors)
```
Comment (agramfort) [inline mne/decoding/csp.py]: why this change?

File: mne/fixes.py [hunk 5]
```diff
@@ -1219,3 +1219,285 @@ def __repr__(self):
     # __getstate__ and __setstate__ are omitted because they only contain
     # conditionals that are not satisfied by our objects (e.g.,
     # ``if type(self).__module__.startswith('sklearn.')``.
+
+
+###############################################################################
+# Copied from sklearn to simplify code paths
+
+def empirical_covariance(X, assume_centered=False):
```
Comment (agramfort) [inline mne/fixes.py]: why this big diff?

File: mne/decoding/csp.py (hunk 4 above)
Comment (larsoner) [reply mne/decoding/csp.py]: `pinv` failed for me on some problem where `pinv2` succeeded

File: mne/fixes.py (hunk 5 above)
Comment (larsoner) [reply mne/fixes.py]: We don't require `sklearn` for emprical, but I unified the code path to always use `sklearn`-like estimators (whereas before we manually implemented the `np.cov` equivalent one) as it simplifies a lot of code. Now we need to copy in the `sklearn` empirical cov estimator to continue not to require `sklearn` when doing `method='empirical'`.

File: mne/decoding/csp.py (hunk 3 above)
Comment (larsoner) [reply mne/decoding/csp.py]: It makes these methods less like the `compute_covariance` equivalent (why have `method_params` passed separately there but not here?) but I'll change it.

Comment (larsoner) [reply mne/decoding/csp.py]: Thinking about it more, removing `method_params` and forcing people to use `partial` will have a couple of negative consequences:

1. The `cov.py` code will need to be changed to handle `callable`, which will (based on its current design) make it more complex / harder to understand.
2. This change will make the `compute_covariance` and `decoding` methods less similar -- a big goal was to unify them as much as possible.

A simpler alternative that avoids these drawbacks is change `method_params` to match the `reg` name -> `reg_params`. Then in `compute_covariance` we have `method` and `method_params`, in decoding we have `reg` and `reg_params`, which seems straightforward enough to me.

**Comment by larsoner** [comment]
@agramfort comments addressed

**Comment by larsoner** [comment]
Squashed and rebased to fix conflict

**Comment by larsoner** [comment]
@agramfort happy with the changes?

File: mne/decoding/csp.py (hunk 3 above)
Comment (agramfort) [reply mne/decoding/csp.py]: I would call it:

cov_method_params

File: mne/decoding/csp.py [hunk 6]
```diff
@@ -680,19 +680,24 @@ class SPoC(CSP):
...
+    log : None | bool (default None)
         If transform_into == 'average_power' and log is None or True, then
         applies a log transform to standardize the features, else the features
         are z-scored. If transform_into == 'csp_space', then log must be None.
     transform_into : {'average_power', 'csp_space'}
         If 'average_power' then self.transform will return the average power of
         each spatial filter. If 'csp_space' self.transform will return the data
         in CSP space. Defaults to 'average_power'.
+    method_params : dict | None
+        Parameters to pass to :func:`mne.compute_covariance`.
```
Comment (agramfort) [inline mne/decoding/csp.py]: cov_method_params

File: mne/fixes.py [hunk 7]
```diff
@@ -1221,6 +1221,287 @@ def __repr__(self):
     # ``if type(self).__module__.startswith('sklearn.')``.
 
 
+###############################################################################
+# Copied from sklearn to simplify code paths
+
+def empirical_covariance(X, assume_centered=False):
+    """Computes the Maximum likelihood covariance estimator
+
+
```
Comment (agramfort) [inline mne/fixes.py]: extra blank line

**Comment by agramfort** [comment]
that's it for me for now. Sorry @larsoner it took me so long.

File: mne/fixes.py [hunk 8]
```diff
@@ -1221,6 +1221,287 @@ def __repr__(self):
...
+            Squared Mahalanobis distances of the observations.
+
+        """
+        precision = self.get_precision()
+        # compute mahalanobis distances
+        centered_obs = observations - self.location_
+        mahalanobis_dist = np.sum(
+            np.dot(centered_obs, precision) * centered_obs, 1)
+
+        return mahalanobis_dist
```
[the remark this answers is not in the document]
Comment (larsoner) [reply mne/fixes.py]: You [asked this in December](https://github.com/mne-tools/mne-python/pull/4808#discussion_r158832433) :)

>> why this big diff?

> We don't require sklearn for emprical, but I unified the code path to always use sklearn-like estimators (whereas before we manually implemented the np.cov equivalent one) as it simplifies a lot of code. Now we need to copy in the sklearn empirical cov estimator to continue not to require sklearn when doing method='empirical'.

File: mne/fixes.py (hunk 7 above)
Comment (larsoner) [reply mne/fixes.py]: This is verbatim copy-paste from `sklearn` so I'd rather leave it

**Comment by larsoner** [comment]
`method_params` changed to `cov_method_params` in `mne/decoding/*`

**Comment by agramfort** [comment]
thanks @larsoner
